luma/core/device.py

```diff
@@ -168,6 +173,90 @@ def display(self, image):
+
+
+class fb(device):
+    """
+    Pseudo-device that acts like a physical display, except that it renders
+    to a Linux framebuffer device at /dev/fbN (where N=0, 1, ...). This is specifically
+    targetted to allow the luma classes to be used on higher-resolution displays that
+    leverage kernel-based display drivers.
+
+    .. note:
+        Currently only supports 16-bit and 24-bit RGB color depths.
+
+    :param device: the Linux framebuffer device (e.g. `/dev/fb0`). If no device
+        is given, the device is determined from the `FRAMEBUFFER` environmental
+        variable instead. See https://www.kernel.org/doc/html/latest/fb/framebuffer.html
+        for more details.
+
+    .. versionadded:: 2.0.0
+    """
+
+    def __init__(self, device=None, **kwargs):
+        super(fb, self).__init__(serial_interface=noop())
+        self.id = self.__get_display_id(device)
+        (width, height) = self.__config("virtual_size")
+        self.bits_per_pixel = next(self.__config("bits_per_pixel"))
+        image_converters = {
+            16: self.__toRGB565,
+            24: self.__toRGB,
+        }
+        assert self.bits_per_pixel in image_converters, f"Unsupported bit-depth: {self.bits_per_pixel}"
+        self.__image_converter = image_converters[self.bits_per_pixel]
+
+        self.capabilities(width, height, rotate=0, mode="RGB")
+
+    def __get_display_id(self, device):
+        """
+        Extract the display-id from the device which is usually referred in
+        the form `/dev/fbN` where N is numeric. If no device is given, defer
+        to the FRAMEBUFFER environmental variable.
+
+        See https://www.kernel.org/doc/html/latest/fb/framebuffer.html for more details.
+        """
+        if device is None:
+            device = os.environ.get("FRAMEBUFFER", "/dev/fb0")
+
+        if device.startswith("/dev/fb"):
+            return int(device[7:])
+
+        raise luma.core.error.DeviceNotFoundError(
+            "Invalid/unsupported framebuffer: {}".format(device)
+        )
+
+    def __config(self, section):
+        path = "/sys/class/graphics/fb{0}/{1}".format(self.id, section)
+        with open(path, "r") as fp:
+            for value in fp.read().strip().split(","):
+                if value:
+                    yield int(value)
+
+    def __toRGB565(self, image):
+        for r, g, b in image.getdata():
+            yield g << 3 & 0xE0 | b >> 3
+            yield r & 0xF8 | g >> 5
+
+    def __toRGB(self, image):
+        return image.tobytes()
+
+    def display(self, image):
+        """
+        Takes a :py:mod:`PIL.Image` and converts it for consumption on the
+        given /dev/fbX framebuffer device.
+
+        :param image: Image to display.
+        :type image: PIL.Image.Image
+        """
+        assert image.mode == self.mode
+        assert image.size == self.size
+
+        image = self.preprocess(image)
+        path = "/dev/fb{}".format(self.id)
```

Comment:
use f-string

Reply:
fixed + updated in a few other places too